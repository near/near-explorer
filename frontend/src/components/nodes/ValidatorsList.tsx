import BN from "bn.js";
import { Component } from "react";

import * as N from "../../libraries/explorer-wamp/nodes";

import ValidatorRow from "./ValidatorRow";

const epochValidatorsStake = new Map();

interface Props {
  validators: any;
  pages: any;
}

class ValidatorsList extends Component<Props> {
  render() {
    const {
      validators,
      pages: { startPage, endPage, activePage, itemsPerPage },
    } = this.props;

    let validatorsList = validators.sort(
      (a: N.ValidationNodeInfo, b: N.ValidationNodeInfo) => {
        // we take "active", "joining", "leaving" validators and sort them firstly
        // after then we sort the rest
        const validatingGroup = ["active", "joining", "leaving"];
        if (
          a.stakingStatus &&
          validatingGroup.indexOf(a.stakingStatus) >= 0 &&
          b.stakingStatus &&
          validatingGroup.indexOf(b.stakingStatus) >= 0
        ) {
          return new BN(b.currentStake).cmp(new BN(a.currentStake));
        } else if (
          (a.proposedStake || b.proposedStake) &&
          a.stakingStatus &&
          validatingGroup.indexOf(a.stakingStatus) < 0 &&
          b.stakingStatus &&
          validatingGroup.indexOf(b.stakingStatus) < 0
        ) {
          return new BN(b.proposedStake || 0).cmp(new BN(a.proposedStake || 0));
        } else if (
          a.stakingStatus &&
          validatingGroup.indexOf(a.stakingStatus) < 0 &&
          b.stakingStatus &&
          validatingGroup.indexOf(b.stakingStatus) < 0
        ) {
          return new BN(b.currentStake).cmp(new BN(a.currentStake));
        }
      }
    );

    // filter validators list by 'active' and 'leaving' validators to calculate cumulative
    // stake only for those validators
    const activeValidatorsList = validatorsList.filter(
      (i: N.ValidationNodeInfo) =>
        i.stakingStatus && ["active", "leaving"].indexOf(i.stakingStatus) >= 0
    );

    const totalStake = activeValidatorsList.reduce(
      (acc: BN, node: N.ValidationNodeInfo) =>
        acc.add(new BN(node.currentStake)),
      new BN(0)
    ) as BN;

    activeValidatorsList.forEach(
      (validator: N.ValidationNodeInfo, index: number) => {
        let total = new BN(0);
        for (let i = 0; i <= index; i++) {
          total = total.add(new BN(activeValidatorsList[i].currentStake));
          epochValidatorsStake.set(validator.account_id, total);
        }
      }
    );

    validatorsList.forEach((validator: N.ValidationNodeInfo, index: number) => {
      const validatorCumStake = epochValidatorsStake.get(validator.account_id);
      if (validatorCumStake) {
        validatorsList[index].cumulativeStakeAmount = validatorCumStake;
      }
    });

    return (
      <>
        {validatorsList
          .slice(startPage - 1, endPage)
          .map((node: N.ValidationNodeInfo, index: number) => (
            <ValidatorRow
              key={node.account_id}
              node={node}
              index={activePage * itemsPerPage + index + 1}
              totalStake={totalStake}
            />
          ))}
      </>
    );
  }
}

export default ValidatorsList;
