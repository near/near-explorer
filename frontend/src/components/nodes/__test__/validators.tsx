import JSBI from "jsbi";
import { ValidatorFullData } from "../../../types/common";
import * as BI from "../../../libraries/bigint";

export const VALIDATORS_LIST: ValidatorFullData[] = [
  {
    accountId: "staked.poolv1.near",
    publicKey: "ed25519:3JBVXqenru2ErAM1kHQ8qfd29dCkURLd6JKrFgtmcDTZ",
    currentEpoch: {
      stake: "42476926077593266003727024545752",
      progress: {
        blocks: {
          produced: 1257,
          total: 1257,
        },
        chunks: {
          produced: 0,
          total: 4,
        },
      },
    },
    nextEpoch: {
      stake: "42585554199314238406961230826241",
    },
    poolInfo: {
      fee: {
        numerator: 10,
        denominator: 100,
      },
      delegatorsCount: 874,
    },
    telemetry: {
      ipAddress: "44.226.195.207",
      nodeId: "ed25519:Gz76mL1D77d1nDZQi1jZM6WwdH8QEAHt3sor2khaFQLs",
      lastSeen: 1622326196323,
      lastHeight: 38639991,
      status: "NoSync",
      agentName: "near-rs",
      agentVersion: "1.19.1",
      agentBuild: "add9c96b",
      latitude: "45.8491",
      longitude: "-119.7143",
      city: "Boardman",
    },
  },
  // 'active' without description
  {
    accountId: "bisontrails.poolv1.near",
    publicKey: "ed25519:Emk6wQJtpQZRJCvvPmmwP9GD2Pk37xxRpmb5uRvJpX62",
    currentEpoch: {
      stake: "20814924218405478221870785686346",
      progress: {
        blocks: {
          produced: 630,
          total: 630,
        },
        chunks: {
          produced: 2,
          total: 6,
        },
      },
    },
    nextEpoch: {
      stake: "20816469293003621239852045782324",
    },
    poolInfo: {
      fee: {
        numerator: 10,
        denominator: 100,
      },
      delegatorsCount: 436,
    },
    telemetry: {
      ipAddress: "34.222.123.106",
      nodeId: "ed25519:2VaBhZM54NukrPenMcCLwmitaKZaSPnyEhqr7mKjDvPj",
      lastSeen: 1622326199959,
      lastHeight: 38639995,
      status: "NoSync",
      agentName: "near-rs",
      agentVersion: "1.19.1",
      agentBuild: "add9c96b",
      latitude: "45.8491",
      longitude: "-119.7143",
      city: "Boardman",
    },
  },
  // 'active' with all data
  {
    accountId: "astro-stakers.poolv1.near",
    publicKey: "ed25519:2nPSBCzjqikgwrqUMcuEVReJhmkC91eqJGPGqH9sZc28",
    currentEpoch: {
      stake: "17776328380244893436869022330562",
      progress: {
        blocks: {
          produced: 525,
          total: 525,
        },
        chunks: {
          produced: 0,
          total: 10,
        },
      },
    },
    nextEpoch: {
      stake: "17777040818619949427934165446355",
    },
    poolInfo: {
      fee: {
        numerator: 1,
        denominator: 100,
      },
      delegatorsCount: 1774,
    },
    telemetry: {
      ipAddress: "145.239.1.147",
      nodeId: "ed25519:FrbiAcqoV4ZrGXaSgf2faXgT6BSa3jRG1NWYvns8Kyho",
      lastSeen: 1622326192345,
      lastHeight: 38639987,
      status: "NoSync",
      agentName: "near-rs",
      agentVersion: "1.19.1",
      agentBuild: "add9c96b",
      latitude: "51.2993",
      longitude: "9.491",
      city: "",
    },
    description: {
      twitter: "@astrostakers",
      url: "astrostakers.com",
      email: "heyheyhey@astrostakers.com",
      description:
        "1% fee in perpetuity secured for NEAR community. Run by Silicon Valley engineers for service permanent availability. We drop rhymes, not blocks",
      countryCode: "AQ",
    },
  },
  // joining
  {
    accountId: "baziliknear.poolv1.near",
    publicKey: "ed25519:E4LAWdgLifBEoaWvhRNy5vpdAnUc3GsUHePeiAurZY5v",
    nextEpoch: {
      stake: "3735553291238507324807526403326",
    },
    afterNextEpoch: {
      stake: "3736122287791333873791529790832",
    },
    poolInfo: {
      fee: {
        numerator: 3,
        denominator: 100,
      },
      delegatorsCount: 148,
    },
    telemetry: {
      ipAddress: "135.181.24.210",
      nodeId: "ed25519:6HBtiS73NrRUCrnKHLzParNBraGaT2NQRX5o51FYaVH9",
      lastSeen: 1622510992841,
      lastHeight: 38825448,
      status: "NoSync",
      agentName: "near-rs",
      agentVersion: "1.19.1",
      agentBuild: "unknown",
      latitude: "43.6319",
      longitude: "-79.3716",
      city: "",
    },
    description: {
      email: "baziliknear@gmail.com",
      description: "pool commission 0% until 01.01.2021, then 1%",
      twitter: "etherscaam",
    },
  },
  // proposal
  {
    accountId: "cryptium.poolv1.near",
    publicKey: "ed25519:5Y9hW8cKBb5RnsJBqttHHC5ujz5zcZZ5xnrJPwkCWmGQ",
    afterNextEpoch: {
      stake: "2656963060607021394774204108064",
    },
    poolInfo: {
      fee: {
        numerator: 5,
        denominator: 100,
      },
      delegatorsCount: 129,
    },
    telemetry: {
      ipAddress: "135.181.179.176",
      nodeId: "ed25519:268mmTfy4oteCbzBY4HusxYZtDZaWdQaiM5HtTbQmtY5",
      lastSeen: 1622326217984,
      lastHeight: 38640013,
      status: "NoSync",
      agentName: "near-rs",
      agentVersion: "1.19.1",
      agentBuild: "add9c96b",
      latitude: "43.6319",
      longitude: "-79.3716",
      city: "",
    },
    description: {
      countryCode: "CH",
      description: "Secure and available validation from the Swiss Alps!",
      email: "hello@cryptium.ch",
      twitter: "CryptiumLabs",
      url: "https://cryptium.ch/",
    },
  },
  // newcomer
  {
    accountId: "huobipool.poolv1.near",
    contractStake: "1200162400050342684350218962960",
    poolInfo: {
      fee: {
        numerator: 10,
        denominator: 100,
      },
      delegatorsCount: 10,
    },
  },
  // idle
  {
    accountId: "ashert.poolv1.near",
    contractStake: "442471954465888931932802773752",
    telemetry: {
      ipAddress: "34.219.20.147",
      nodeId: "ed25519:5tFfDk4r5mbBtxjY1x2v4XcFnbcW6Lm8AuJvzbhcyCBS",
      lastSeen: 1622326216916,
      lastHeight: 38640012,
      status: "NoSync",
      agentName: "near-rs",
      agentVersion: "1.19.1",
      agentBuild: "add9c96b",
      latitude: "45.8491",
      longitude: "-119.7143",
      city: "Boardman",
    },
    poolInfo: {
      fee: {
        numerator: 7,
        denominator: 100,
      },
      delegatorsCount: 9,
    },
  },
  // idle with minimum data
  {
    accountId: "sphere.poolv1.near",
    contractStake: "50000867090131772400000000",
    poolInfo: {
      fee: {
        numerator: 8,
        denominator: 100,
      },
      delegatorsCount: 1,
    },
  },
  // leaving
  {
    accountId: "01node.poolv1.near",
    publicKey: "ed25519:5xz7EbcnPqabwoFezdJBxieK8S7XLsdHHuLwM4vLLhFt",
    currentEpoch: {
      stake: "3632096312066963778697815591239",
      progress: {
        blocks: {
          produced: 105,
          total: 105,
        },
        chunks: {
          produced: 1,
          total: 3,
        },
      },
    },
    poolInfo: {
      fee: {
        numerator: 1,
        denominator: 100,
      },
      delegatorsCount: 512,
    },
    telemetry: {
      ipAddress: "91.207.102.230",
      nodeId: "ed25519:Bjy4baS4MQM7FEQS4YS73boS4hJ6sdEgrfZCjqS5SXoo",
      lastSeen: 1622326197935,
      lastHeight: 38639993,
      status: "NoSync",
      agentName: "near-rs",
      agentVersion: "1.19.1",
      agentBuild: "add9c96b",
      latitude: "46",
      longitude: "25",
      city: "",
    },
    description: {
      countryCode: "ro",
      email: "hello@01node.com",
      description:
        "01node Professional Staking Services for Near, Skale, Solana, Cosmos, Iris",
      twitter: "@01node",
      url: "https://01node.com/",
    },
  },
];

export const getCumulativeStake = (node: ValidatorFullData): JSBI => {
  let cumulativeStake = BI.zero;
  for (const validator of VALIDATORS_LIST) {
    cumulativeStake = JSBI.add(
      cumulativeStake,
      JSBI.BigInt(validator.currentEpoch?.stake || "0")
    );
    if (validator === node) {
      return cumulativeStake;
    }
  }
  return cumulativeStake;
};
