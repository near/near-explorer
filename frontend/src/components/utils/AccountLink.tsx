import Link from "next/link";

export interface Props {
  accountId: string;
  children?: React.ReactNode;
}

export default ({ accountId, children }: Props) => (
  <Link href="/accounts/[id]" as={`/accounts/${accountId}`}>
    <a className="account-link">{children || `@${accountId}`}</a>
  </Link>
);
