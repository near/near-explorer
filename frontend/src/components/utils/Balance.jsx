/// Copied from near-wallet project:
/// https://github.com/nearprotocol/near-wallet/blob/41cb65246134308dd553b532dfb314b45b38b65c/src/components/common/Balance.js

import { utils } from "nearlib";
import { OverlayTrigger, Tooltip } from "react-bootstrap";

const Balance = ({ amount }) => {
  if (!amount) {
    throw new Error("amount property should not be null");
  }
  let amountShow = convertToShow(amount);
  let amountPrecise = formatPreciseNEAR(amount);
  return (
    <OverlayTrigger
      placement={"bottom"}
      overlay={<Tooltip>{amountPrecise} yoctoⓃ</Tooltip>}
    >
      <span>{amountShow} Ⓝ</span>
    </OverlayTrigger>
  );
};

const convertToShow = amount => {
  if (amount === "0") {
    return "0";
  }
  return formatNEAR(amount);
};

const formatNEAR = amount => {
  let ret = utils.format.formatNearAmount(amount, 5);
  if (ret === "0" && amount > 0) {
    return "<0.00001";
  }
  return ret;
};

const formatPreciseNEAR = amount => {
  const REG = /(?=(\B)(\d{3})+$)/g;
  return amount.replace(REG, ",");
};
export default Balance;
