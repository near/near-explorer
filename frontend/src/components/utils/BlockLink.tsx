import Link from "next/link";

export interface Props {
  blockHash: string;
  children?: React.ReactNode;
}

export default ({ blockHash, children }: Props) => (
  <Link href="/blocks/[hash]" as={`/blocks/${blockHash}`}>
    <a className="block-link">
      {children || `${blockHash.substring(0, 7)}...`}
    </a>
  </Link>
);
