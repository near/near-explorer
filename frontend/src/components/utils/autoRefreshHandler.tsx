import InfiniteScroll from "react-infinite-scroll-component";

import React from "react";

import PaginationSpinner from "./PaginationSpinner";

interface Config {
  fetchDataFn: Function;
  count: number;
  category: string;
  dashboard?: boolean;
}

export default (
  WrappedComponent: React.ComponentType<any>,
  config: Config,
  props?: any
) => {
  return class extends React.Component {
    constructor(props: any) {
      super(props);
      this.timer = null;
    }

    timer: ReturnType<typeof setTimeout> | null;

    state = {
      items: Array<any>(),
      itemsLength: config.count,
      display: false,
      hasMore: true,
      loading: false
    };

    componentDidMount() {
      this.timer = setTimeout(this.regularFetchInfo, 0);
    }

    componentWillUnmount() {
      const timer = this.timer;
      this.timer = null;
      if (timer !== null) {
        clearTimeout(timer);
      }
    }

    componentDidUpdate(preProps: any) {
      if (this.props !== preProps) {
        this.setState({
          items: Array<any>(),
          itemsLength: config.count,
          display: false,
          hasMore: true,
          loading: false
        });
      }
    }

    regularFetchInfo = () => {
      if (config.dashboard) {
        this.setState({ hasMore: false });
      }
      if (this.state.itemsLength > 0) {
        this.fetchInfo(this.state.itemsLength);
        if (this.timer !== null) {
          this.timer = setTimeout(this.regularFetchInfo, 10000);
        }
      }
    };

    fetchInfo = (count: number) => {
      config
        .fetchDataFn(count)
        .then((items: any) => {
          let newState: any;
          if (items.length > 0) {
            newState = {
              items,
              itemsLength: items.length,
              display: true
            };
            if (items.length < config.count) {
              newState.hasMore = false;
            }
          } else {
            newState = {
              hasMore: false,
              display: true,
              itemsLength: 0
            };
          }
          this.setState(newState);
        })
        .catch((err: Error) => {
          console.error(err);
        });
    };

    fetchMoreData = async () => {
      this.setState({ loading: true });
      const endTimestamp = this.getEndTimestamp(config.category);
      config
        .fetchDataFn(config.count, endTimestamp)
        .then((newData: any) => {
          let newState: any;
          if (newData.length > 0) {
            const items = this.state.items.concat(newData);
            newState = {
              items,
              itemsLength: items.length,
              loading: false
            };
            if (newData.length < config.count) {
              newState.hasMore = false;
            }
          } else {
            newState = { hasMore: false, loading: false };
          }
          this.setState(newState);
        })
        .catch((err: Error) => console.error(err));
    };

    getEndTimestamp = (category: string) => {
      let endTimestamp;
      switch (category) {
        case "Account":
          endTimestamp = this.state.items[this.state.items.length - 1]
            .createdAtBlockTimestamp;
          break;
        case "Block":
          endTimestamp = this.state.items[this.state.items.length - 1]
            .timestamp;
          break;
        case "Node":
          endTimestamp = this.state.items[this.state.items.length - 1].lastSeen;
          break;
        case "Transaction":
          endTimestamp = this.state.items[this.state.items.length - 1]
            .blockTimestamp;
          break;
        default:
          endTimestamp = undefined;
      }
      return endTimestamp;
    };

    render() {
      if (!this.state.display) {
        return <PaginationSpinner hidden={false} />;
      }
      return (
        <>
          <InfiniteScroll
            dataLength={this.state.items.length}
            next={this.fetchMoreData}
            hasMore={this.state.hasMore}
            loader={
              this.state.loading ? (
                <PaginationSpinner hidden={false} />
              ) : (
                <>
                  <button
                    onClick={this.fetchMoreData}
                    className="load-button"
                    style={{ display: this.state.hasMore ? "block" : "none" }}
                  >
                    Load More
                  </button>
                </>
              )
            }
            style={{ overflowX: "hidden" }}
          >
            <WrappedComponent items={this.state.items} {...props} />
          </InfiniteScroll>
          <style jsx global>{`
            .load-button {
              width: 100px;
              background-color: #f8f8f8;
              display: block;
              text-align: center;
              text-decoration: none;
              font-family: BentonSans;
              font-size: 14px;
              color: #0072ce;
              font-weight: bold;
              text-transform: uppercase;
              margin: 20px auto;
              border-radius: 30px;
              padding: 8px 0;
              cursor: pointer;
              border: none;
            }
          `}</style>
        </>
      );
    }
  };
};
