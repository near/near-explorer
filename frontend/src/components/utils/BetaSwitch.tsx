import React from "react";

import { Switch } from "@/frontend/components/utils/Switch";
import { useBetaOptions } from "@/frontend/hooks/use-beta-options";
import { styled } from "@/frontend/libraries/styles";

const Wrapper = styled("div", {
  display: "flex",
  alignItems: "center",
  color: "#ccc",

  "& > * + *": {
    marginLeft: 8,
  },

  variants: {
    enabled: {
      true: {
        color: "#0072ce",
      },
    },
  },
});

export const BetaSwitch: React.FC = () => {
  const [betaOptions, setBetaOptions] = useBetaOptions();
  return (
    <Wrapper enabled={betaOptions?.enabled}>
      <span>Beta</span>
      <Switch
        checked={Boolean(betaOptions?.enabled)}
        onChange={(enabled) => setBetaOptions((prev) => ({ ...prev, enabled }))}
      />
    </Wrapper>
  );
};
