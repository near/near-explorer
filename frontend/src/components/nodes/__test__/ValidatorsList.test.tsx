import { renderElement } from "../../../testing/utils";

import ValidatorsList from "../ValidatorsList";

import { getCumulativeStake, VALIDATORS_LIST } from "./validators";

const totalStake = getCumulativeStake(
  VALIDATORS_LIST[VALIDATORS_LIST.length - 1]
);

describe("<ValidatorsList />", () => {
  it("renders validators list", () => {
    expect(
      renderElement(
        <ValidatorsList
          validators={VALIDATORS_LIST.slice(0, 5)}
          selectedPageIndex={0}
          totalStake={totalStake.toString()}
        />
      )
    ).toMatchSnapshot();
  });
});
