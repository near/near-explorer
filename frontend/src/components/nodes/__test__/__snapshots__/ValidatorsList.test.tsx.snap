// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ValidatorsList /> renders validators list 1`] = `
Array [
  <tr
    className="jsx-4079813474 table-row validator-nodes-row mx-0 "
  >
    <td
      className="collapse-row-arrow"
      onClick={[Function]}
    >
      <img
        src="/static/images/icon-maximize.svg"
        style={
          Object {
            "width": "16px",
          }
        }
      />
    </td>
    <td
      className="order"
    >
      25
    </td>
    <td
      className="country-flag"
    >
      <img
        onBlur={[Function]}
        onFocus={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        src="https://cdnjs.cloudflare.com/ajax/libs/flag-icon-css/3.4.3/flags/4x3/us.svg"
        style={
          Object {
            "display": "inline-block",
            "height": "1em",
            "lineHeight": "14px",
            "verticalAlign": "middle",
            "width": "22px",
          }
        }
      />
    </td>
    <td>
      <div
        className="align-items-center row no-gutters"
      >
        <div
          className="validators-node-label col-2"
        >
          <span
            className="validating-label joining badge"
            onBlur={[Function]}
            onFocus={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
          >
            Joining
          </span>
        </div>
        <div
          className="validator-name col"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="validator-nodes-text col"
              title="@astro-stakers.poolv1.near"
            >
              astro-stakers.poolv1.near
            </div>
          </div>
          <div
            className="row no-gutters"
          >
            <div
              className="validator-nodes-text validator-node-pub-key col"
              title="ed25519:2nPSBCzjqikgwrqUMcuEVReJhmkC91eqJGPGqH9sZc28"
            >
              ed25519:2nPSBCzjqikgwrqUMcuEVReJhmkC91eqJGPGqH9sZc28
            </div>
          </div>
        </div>
      </div>
    </td>
    <td>
      1%
    </td>
    <td>
      <div
        className="spinner-border spinner-border-sm"
      />
    </td>
    <td
      className="text-right validator-nodes-text stake-text"
    >
      <span
        onBlur={[Function]}
        onFocus={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
      >
        19,407,280
         
        NEAR
      </span>
      <br />
      <small>
        -
        <span
          onBlur={[Function]}
          onFocus={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
        >
          15,708,597
           
          NEAR
        </span>
      </small>
    </td>
    <td>
      <div
        className="jsx-1785419880 cumulative-stake-chart"
      >
        <div
          className="jsx-1785419880 total-value"
          style={
            Object {
              "width": "-250.93%",
            }
          }
        />
        <div
          className="jsx-1785419880 current-value"
          style={
            Object {
              "width": "0%",
            }
          }
        />
        <div
          className="jsx-1785419880 cumulative-stake-label"
        >
          N/A
        </div>
      </div>
    </td>
  </tr>,
  <tr
    className="jsx-4079813474 table-row validator-nodes-row mx-0 "
  >
    <td
      className="collapse-row-arrow"
      onClick={[Function]}
    >
      <img
        src="/static/images/icon-maximize.svg"
        style={
          Object {
            "width": "16px",
          }
        }
      />
    </td>
    <td
      className="order"
    >
      26
    </td>
    <td
      className="country-flag"
    >
      <div
        style={
          Object {
            "backgroundColor": "#f0f0f1",
            "color": "#72727a",
            "fontSize": "8px",
            "lineHeight": "16px",
            "textAlign": "center",
            "width": "22px",
          }
        }
      >
        ?
      </div>
    </td>
    <td>
      <div
        className="align-items-center row no-gutters"
      >
        <div
          className="validators-node-label col-2"
        >
          <span
            className="validating-label leaving badge"
            onBlur={[Function]}
            onFocus={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
          >
            Kickout
          </span>
        </div>
        <div
          className="validator-name col"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="validator-nodes-text col"
              title="@baziliknear.poolv1.near"
            >
              baziliknear.poolv1.near
            </div>
          </div>
          <div
            className="row no-gutters"
          >
            <div
              className="validator-nodes-text validator-node-pub-key col"
              title="ed25519:E4LAWdgLifBEoaWvhRNy5vpdAnUc3GsUHePeiAurZY5v"
            >
              ed25519:E4LAWdgLifBEoaWvhRNy5vpdAnUc3GsUHePeiAurZY5v
            </div>
          </div>
        </div>
      </div>
    </td>
    <td>
      3%
    </td>
    <td>
      <div
        className="spinner-border spinner-border-sm"
      />
    </td>
    <td
      className="text-right validator-nodes-text stake-text"
    >
      <span
        onBlur={[Function]}
        onFocus={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
      >
        4,035,443
         
        NEAR
      </span>
      <br />
      <small>
        +
        <span
          onBlur={[Function]}
          onFocus={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
        >
          756,812
           
          NEAR
        </span>
      </small>
    </td>
    <td>
      <div
        className="jsx-1785419880 cumulative-stake-chart"
      >
        <div
          className="jsx-1785419880 total-value"
          style={
            Object {
              "width": "0%",
            }
          }
        />
        <div
          className="jsx-1785419880 current-value"
          style={
            Object {
              "width": "52.17%",
            }
          }
        />
        <div
          className="jsx-1785419880 cumulative-stake-label"
        >
          52.17%
        </div>
      </div>
    </td>
  </tr>,
  <tr
    className="jsx-3940587247 cumulative-stake-holders-row"
  >
    <td
      className="jsx-3940587247 warning-text text-center"
      colSpan={8}
    >
      Validators 1 - 26 hold a cumulative stake above 33%. Delegating to the validators below improves the decentralization of the network.
    </td>
  </tr>,
  <tr
    className="jsx-4079813474 table-row validator-nodes-row mx-0 "
  >
    <td
      className="collapse-row-arrow"
      onClick={[Function]}
    >
      <img
        src="/static/images/icon-maximize.svg"
        style={
          Object {
            "width": "16px",
          }
        }
      />
    </td>
    <td
      className="order"
    >
      27
    </td>
    <td
      className="country-flag"
    >
      <img
        onBlur={[Function]}
        onFocus={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        src="https://cdnjs.cloudflare.com/ajax/libs/flag-icon-css/3.4.3/flags/4x3/us.svg"
        style={
          Object {
            "display": "inline-block",
            "height": "1em",
            "lineHeight": "14px",
            "verticalAlign": "middle",
            "width": "22px",
          }
        }
      />
    </td>
    <td>
      <div
        className="align-items-center row no-gutters"
      >
        <div
          className="validators-node-label col-2"
        >
          <span
            className="validating-label active badge"
            onBlur={[Function]}
            onFocus={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
          >
            Active
          </span>
        </div>
        <div
          className="validator-name col"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="validator-nodes-text col"
              title="@audit_one.poolv1.near"
            >
              audit_one.poolv1.near
            </div>
          </div>
          <div
            className="row no-gutters"
          >
            <div
              className="validator-nodes-text validator-node-pub-key col"
              title="ed25519:9x2LqdGVL8MKbAziezMhQjPCKKeFAaExeyJpVvBvGuAC"
            >
              ed25519:9x2LqdGVL8MKbAziezMhQjPCKKeFAaExeyJpVvBvGuAC
            </div>
          </div>
        </div>
      </div>
    </td>
    <td>
      7%
    </td>
    <td>
      <div
        className="spinner-border spinner-border-sm"
      />
    </td>
    <td
      className="text-right validator-nodes-text stake-text"
    >
      <span
        onBlur={[Function]}
        onFocus={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
      >
        3,698,683
         
        NEAR
      </span>
      <br />
      <small>
        +
        <span
          onBlur={[Function]}
          onFocus={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
        >
          1,093,573
           
          NEAR
        </span>
      </small>
    </td>
    <td>
      <div
        className="jsx-1785419880 cumulative-stake-chart"
      >
        <div
          className="jsx-1785419880 total-value"
          style={
            Object {
              "width": "52.18%",
            }
          }
        />
        <div
          className="jsx-1785419880 current-value"
          style={
            Object {
              "width": "47.82%",
            }
          }
        />
        <div
          className="jsx-1785419880 cumulative-stake-label"
        >
          100%
        </div>
      </div>
    </td>
  </tr>,
  <tr
    className="jsx-4079813474 table-row validator-nodes-row mx-0 "
  >
    <td
      className="collapse-row-arrow"
      onClick={[Function]}
    >
      <img
        src="/static/images/icon-maximize.svg"
        style={
          Object {
            "width": "16px",
          }
        }
      />
    </td>
    <td
      className="order"
    >
      28
    </td>
    <td
      className="country-flag"
    >
      <img
        onBlur={[Function]}
        onFocus={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        src="https://cdnjs.cloudflare.com/ajax/libs/flag-icon-css/3.4.3/flags/4x3/ru.svg"
        style={
          Object {
            "display": "inline-block",
            "height": "1em",
            "lineHeight": "14px",
            "verticalAlign": "middle",
            "width": "22px",
          }
        }
      />
    </td>
    <td>
      <div
        className="align-items-center row no-gutters"
      >
        <div
          className="validators-node-label col-2"
        >
          <span
            className="validating-label proposal badge"
            onBlur={[Function]}
            onFocus={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
          >
            Proposal
          </span>
        </div>
        <div
          className="validator-name col"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="validator-nodes-text col"
              title="@genesislab.poolv1.near"
            >
              genesislab.poolv1.near
            </div>
          </div>
        </div>
      </div>
    </td>
    <td>
      4%
    </td>
    <td>
      <div
        className="spinner-border spinner-border-sm"
      />
    </td>
    <td
      className="text-right validator-nodes-text stake-text"
    >
      <span
        onBlur={[Function]}
        onFocus={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
      >
        4,035,443
         
        NEAR
      </span>
      <br />
      <small>
        +
        <span
          onBlur={[Function]}
          onFocus={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
        >
          756,812
           
          NEAR
        </span>
      </small>
    </td>
    <td>
      <div
        className="jsx-1785419880 cumulative-stake-chart"
      >
        <div
          className="jsx-1785419880 total-value"
          style={
            Object {
              "width": "-52.17%",
            }
          }
        />
        <div
          className="jsx-1785419880 current-value"
          style={
            Object {
              "width": "0%",
            }
          }
        />
        <div
          className="jsx-1785419880 cumulative-stake-label"
        >
          N/A
        </div>
      </div>
    </td>
  </tr>,
  <tr
    className="jsx-4079813474 table-row validator-nodes-row mx-0 "
  >
    <td
      className="collapse-row-arrow"
      onClick={[Function]}
    >
      <img
        src="/static/images/icon-maximize.svg"
        style={
          Object {
            "width": "16px",
          }
        }
      />
    </td>
    <td
      className="order"
    >
      29
    </td>
    <td
      className="country-flag"
    >
      <div
        style={
          Object {
            "backgroundColor": "#f0f0f1",
            "color": "#72727a",
            "fontSize": "8px",
            "lineHeight": "16px",
            "textAlign": "center",
            "width": "22px",
          }
        }
      >
        ?
      </div>
    </td>
    <td>
      <div
        className="align-items-center row no-gutters"
      >
        <div
          className="validators-node-label col-2"
        >
          <span
            className="validating-label on-hold badge"
            onBlur={[Function]}
            onFocus={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
          >
            On hold
          </span>
        </div>
        <div
          className="validator-name col"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="validator-nodes-text col"
              title="@certusone.poolv1.near"
            >
              certusone.poolv1.near
            </div>
          </div>
          <div
            className="row no-gutters"
          >
            <div
              className="validator-nodes-text validator-node-pub-key col"
              title="ed25519:4kwUVNSZ61BT3xbc1fL718UP2RqgvMa5aAYBaoav1wvk"
            >
              ed25519:4kwUVNSZ61BT3xbc1fL718UP2RqgvMa5aAYBaoav1wvk
            </div>
          </div>
        </div>
      </div>
    </td>
    <td>
      10%
    </td>
    <td>
      70
    </td>
    <td
      className="text-right validator-nodes-text stake-text"
    >
      <span
        onBlur={[Function]}
        onFocus={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
      >
        3,714,537
         
        NEAR
      </span>
      <br />
      <small>
        -
        <span
          onBlur={[Function]}
          onFocus={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
        >
          4,408
           
          NEAR
        </span>
      </small>
    </td>
    <td>
      <div
        className="jsx-1785419880 cumulative-stake-chart"
      >
        <div
          className="jsx-1785419880 total-value"
          style={
            Object {
              "width": "-48.02%",
            }
          }
        />
        <div
          className="jsx-1785419880 current-value"
          style={
            Object {
              "width": "0%",
            }
          }
        />
        <div
          className="jsx-1785419880 cumulative-stake-label"
        >
          N/A
        </div>
      </div>
    </td>
  </tr>,
]
`;
